pkg/bmc/credentials.go

```diff
@@ -1,37 +1,190 @@
 package bmc
 
-const (
-	// MissingCredentialsMsg is returned as a validation failure
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	metalkubev1alpha1 "github.com/metalkube/baremetal-operator/pkg/apis/metalkube/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Error is the error type usually returned by functions in the bmc
+// package. It describes the host, whether or not this is a fatal error,
```

Comment:
I'm going to roll up the replies to both @zaneb and @mhrivnak into one reply.

First, let me apologize for some of the dangling bits like `Error.getErr()` that @mhrivnak pointed out.  I rewrote this several times, starting with a few lines of change, to breaking apart `getValidBMCCredentials` into smaller functions before finally landing here and wanted to stop for a moment and see what people thought about this general direction and didn't clean everything up before pausing for some feedback.  I do agree this is now going far beyond the scope of the original PR and should likely be spun into its own thing once there is a clear path.

In any event, at a high level, my primary motivation was here was there seemed to be specific courses of action that need to be taken depending on how BMC retrieval *and* validation behaves.  These have to be handled within `Reconcile` because those paths involve things like setting the error state on hosts, publishing messages, and things generally that do not belong in the isolated modules like `bmc`.  This means that `Reconcile` has to be armed with the precise modes of failure in creating said credentials to respond appropriately and this is where I saw the need to do proper error typing and propagation. Now that said, @mhrivnak makes a fair point that this now dirties the previously simple `bmc` package a bit with some Kubernetes logic and a client.  Given how simple `Credentials` was, really adding anything there would ruin its pristine condition! But alas, I admit I also was not fond of introducing the k8s client there as I was typing it and it also resulted in some copy-and-paste testing boiler plate (e.g. `newHost`) which also isn't good.

However, at the end of the day I think the question we all want to simply answer is "Does this host have sane BMC data" and I started to find that the validation for this question seems split in several places.  This is what I was really setting out to resolve. There is no one place I can go and see how we say yes or no to this question.  Instead, some validation is done within `Reconcile` like [here](https://github.com/metalkube/baremetal-operator/blob/master/pkg/controller/baremetalhost/baremetalhost_controller.go#L202-L219) and [here](https://github.com/metalkube/baremetal-operator/blob/master/pkg/controller/baremetalhost/baremetalhost_controller.go#L220-L237).  While the validation is done here, this pulls error messages from the `bmc` package to surface them.  Then within `bmc`, we will validate the innards of the credentials [here](https://github.com/metalkube/baremetal-operator/blob/master/pkg/bmc/credentials.go#L29-L37).  Right or wrong, I did want to provide some justification for muddying the `bmc` package with a k8s client and all that entails - it allowed all of this to be put into one place and maybe could be a path for getting to single place that can answer the question of "Does this host have sane BMC data." and part of answering that question is talking to Kubernetes to resolve the secret reference if there is a desire to unify it all into one place.  The alternative is heavier `Reconcile` logic and trivial `Credentials` validation like it was.  At what point to `Reconcile` implode under its own weight?

I also used this as an opportunity to re-read the [Errors](https://golang.org/doc/effective_go.html#errors) section of effective Go.  That's mostly what is under the hood here plus a few embellishments I got carried away with that we can ditch.  The thought I had while writing this -- right or wrong -- is that these errors could contain enough useful properties to also allow non-Reconcile users to benefit from them as well, even though none exist today.  I do like this pattern of embedded Err objects on a high level simplistic `bmc.Error` (or whatever it ends up being called) because it means I can create specific errors for specific events and switch on them if ever the need is there.

@zaneb Thanks for the feedback.  No disagreements on anything you have said.  I am not confident the `bmc.Error` struct is exactly where it needs to be in terms of structure (@mhrivnak points out several concerns) but needed to take a moment and get a general sense of feedback on this direction before continuing any further. 

Unless there is a further course correction in feedback, for the moment I will remove the k8s client introduction into credentials, simplify the `bmc.Error` structure and see what everyone thinks then.  I'll also rename the pull request to deal with the fact that this hasn't quite stuck with the original scope.  I can't help myself when I start poking.

Reply:
I need to rewrite some of this.  Fatal is a bit misleading.  An error can either doNotReque or not if I understand the use cases today.